### ISSUE TITLE
Feature request: minimum character count for realText

### ISSUE THREAD

**Issue by FatBoyXPC** [FakerPHP/Faker#37]
I recently wrote a simple loop to ensure `realText()` would provide a minimum character count. It would be nice if this could be added as a parameter to the method itself instead. I'm more than happy to PR this in, and will likely do so over the weekend.

**Comment by pimjansen** [comment]
If im correct it always tries to fill the lenght specified right?

What we should not do is set a minimum only because it can output really really long texts and some locale

**Comment by FatBoyXPC** [comment]
I don't know what "always tries" means. I can tell you that `realText(100)` _might_ give you a 100 character string, might not. I can tell you it will give you a string that's 100 or fewer characters, though. I'm wanting to create a minimum as I had to write a loop in a test recently to accomplish this.

**Comment by pimjansen** [comment]
> I don't know what "always tries" means. I can tell you that `realText(100)` _might_ give you a 100 character string, might not. I can tell you it will give you a string that's 100 or fewer characters, though. I'm wanting to create a minimum as I had to write a loop in a test recently to accomplish this.

The reason it can give you less is because it creates text and it is pretty hard to create valid text since we have to deal with words itself.

**Comment by FatBoyXPC** [comment]
Sure, I understand _why_, but that doesn't change the fact it would be nice to have a _minimum_. Perhaps a way forward is to ensure min/max are different enough by a percentage?

**Comment by pimjansen** [comment]
> Sure, I understand _why_, but that doesn't change the fact it would be nice to have a _minimum_. Perhaps a way forward is to ensure min/max are different enough by a percentage?

Yeah fine by me though and we can do it as a non breaking change i assume.


But the fact is that almost everyone wants to output just a text with around x chars. Im not a fan on only setting the minimum since it will output hundred lines of text all the time.

But setting a max 1000 and min of 900 will still output you around 995 words maybe. Where a min 900 only will give your 2000+ words.

So the only way here is that we add a new method for this to get a minimum length but imo if feels a bit duplicate.

@localheinz what is your view on this?

**Comment by FatBoyXPC** [comment]
Any reason we can't add a parameter to the `realText()` method? I'm not sure I see a problem of setting a 900 minimum and getting a 2000 back?

Also, the min/max I'm talking about is characters, not words.

**Comment by pimjansen** [comment]
> Any reason we can't add a parameter to the `realText()` method? I'm not sure I see a problem of setting a 900 minimum and getting a 2000 back?
> 
> Also, the min/max I'm talking about is characters, not words.

The method is designed to give you proper fake text and not cut-off words. Therefor you could just as easy just create random letters in a sequence.

Adding a min length will counter a max length on the method which is odd to say the least.

What you could do is add a method with a min which defaults to 0 though but that would still mean that you have to call it with 2 params:

Max: 200
Min: 150

And if min and max are too close together it will also not work. It will try to fill up the space with words from the text dictionary till it fills the max count

**Comment by bramceulemans** [comment]
This feels redundant, since `realText` will try to get as close as possible to the `maxNbChars`, you might get unlucky and get a text that is +/- 10 chars smaller, because a long word didn't fit.

**Comment by FatBoyXPC** [comment]
I'm not aware of any 1000 character word in the English language, and here's the loop we ended up writing:

```php
                $longNote = '';
                while (strlen($longNote) <= 1000) {
                        $longNote .= $this->faker->realText(2000, 3);
                }
```

If what you said were true, this loop wouldn't have been necessary.

**Comment by pimjansen** [comment]
> I'm not aware of any 1000 character word in the English language, and here's the loop we ended up writing:
> 
> ```php
>                 $longNote = '';
>                 while (strlen($longNote) <= 1000) {
>                         $longNote .= $this->faker->realText(2000, 3);
>                 }
> ```
> 
> If what you said were true, this loop wouldn't have been necessary.

It does exactly as intended. It will max out the space with a valid text. So it the max is 1000 it will try to fill it as close to 1000. Technically that will indeed mean it are less chars then 1000 due to 2 reasons

- it cant exceed the max
- it should contain valid text words and not random letters

**Comment by FatBoyXPC** [comment]
> So it the max is 1000 it will try to fill it as close to 1000.

But it's not: `$longNote .= $this->faker->realText(2000, 3);`

**Comment by pimjansen** [comment]
> > So it the max is 1000 it will try to fill it as close to 1000.
> 
> But it's not: `$longNote .= $this->faker->realText(2000, 3);`

Correct, with the index size it will try to fill the space with an index of words to keep the text integrity in place as good as possible. So the bigger the index size the more real a text would but but the more offset it will maybe have from the size.

**Comment by FatBoyXPC** [comment]
Regardless, the current implementation makes it possible to only get 1000 words instead of 2000 when 2000 is specified. The implementation should be changed to actually work like you first suggested (get as close to without going over), or we should allow a minimum to be set.

**Comment by pimjansen** [comment]
Closing this issue where the method does what it says. It will output a realText with a `$maxNbChars`.

**Comment by FatBoyXPC** [comment]
I'm honestly floored at this interaction: I showed you code that this doesn't work how you say it works, and now you're closing this issue pretending it does? I realize this is open source and what not, and if you were to say something like "We don't want to support this feature" - I get it. Instead, you're pretending that something works a way that it doesn't.

**Comment by pimjansen** [comment]
> I'm honestly floored at this interaction: I showed you code that this doesn't work how you say it works, and now you're closing this issue pretending it does? I realize this is open source and what not, and if you were to say something like "We don't want to support this feature" - I get it. Instead, you're pretending that something works a way that it doesn't.

I tried to reproduce a few times and at all the tests it returned less then the maximum allowed chars.

@bramceulemans can you maybe run it as well?

**Comment by bramceulemans** [comment]
I ran a few tests as we'll:

**Comment by FatBoyXPC** [comment]
> I tried to reproduce a few times and at all the tests it returned less then the maximum allowed chars

Well that explains it - I'm not logging a bug, I'm asking for a feature request. I'd like for `realText()` to also allow a "minimum" character count, as it doesn't always provide as many as you want. While I understand needing a certain percentage below, 50% seems like quite a stretch to allow.

The test that @bramceulemans ran isn't quite on the same level with the code I found that needed the loop. My team _wanted_ to write `realText(2000)` and get more than 1000 characters back.

Maybe test with the same arguments from my code snippet above? `realText(2000, 3)`

**Comment by bramceulemans** [comment]
The thing is, it will always get as close as possible to the given maximum from what I'm reading, in my testing it was never off more than 20 chars?

**Comment by pimjansen** [comment]
Correct as intended. Mixing a min chars with a max chars does not make any sense. The method will try to return a best suitable matching real text. Based on texts provided by locales it will render it with the index offset.

The method was never build to set up a minimum length.

I also doubting if we even should do stuff like this in core though but those are just my 2 cents.

Feel free to propose a solution but i do not see this happening without running any breaking changes

**Comment by krsriq** [comment]
I'm not talking about adding another `minChars` parameter to `realText`, I really think that there's a bug here.

Executing 
```php
$faker = \Faker\Factory::create();
$faker->seed(252);
$text = $faker->realText(2000, 2);
```

outputs `THE.`, which defintely is no real text. 
I'm pretty sure that there's a way to fix this (without breaking changes), which would incidentally also fix the problem this ticket is about.

But I agree that it's a good idea to think about whether this should even be in the core :)

**Comment by krsriq** [comment]
Created a PR now that doesn't add the requested additional parameter to `realText`, but makes sure that the generated text is at least 80% of the length requested via `$maxNbChars`, so 

```php
$longNote = '';
while (strlen($longNote) <= 1000) {
          $longNote .= $this->faker->realText(2000, 3);
}
```

would not be necessary any more, while the bug uncovered through this discussion is fixed.

**Comment by FatBoyXPC** [comment]
> Mixing a min chars with a max chars does not make any sense.

I would argue that many people would absolutely want "give me real text that's at least much, but under this much".

> The method was never build to set up a minimum length.

I don't think that the implementation should drive the use case. I think it's quite opposite. If the idea of `realText()` is "get as close to the amount of given characters as possible from a sample of real world text" then this clearly has a bug. I also don't think it's a stretch to also say "give me at least an amount of characters, but not more than an amount, from real world text".

**Comment by krsriq** [comment]
Updated the PR now because I realized that it is trivial to provide the functionality requested by @FatBoyXPC while fixing the issue discovered here (`realText` that is only 4 chars long despite `$minNbChars` being 2000).

So while with #152 `realText` should now work in most cases (because it tries to generate text at least the length of 80% of `$maxNbChars`), I've added `Text::realTextBetween($minNbChars = 160, $maxNbChars = 200, $indexSize = 2)` for users that really need to be sure that the generated text is e.g. between 1990 and 2000 chars.
